### ISSUE TITLE
Support Python builtins on iterators in JIT

### ISSUE THREAD

**Issue by narfanar** [pytorch/pytorch#31772]
## 🚀 Feature
The exact example I stumbled upon, but there must be others:

```py
x: List[Tensor]
xrev = list(reversed(x)) # Fails because `reversed` expects only a Tensor in TorchScript
```

## Motivation

Lists of Tensors are somewhat supported in TorchScript and they do have their uses. But some builtins currently do not work with them, like `reversed`.

Plain Tensors cannot replace lists in every way (for example, what if I really don't want to allocate just to get all the elements in a list in one Tensor? What if the elements don't even fit by dimensionality?) so this's missing functionality.

## Pitch

For supported bulitins that support iterators, support List and Tuple values just as Tensors are supported.

## Alternatives

Not sure.

## Additional context

#18627

**Comment by suo** [comment]
Thanks for the report! This is indeed a gap in our coverage for lists, cc @driazati who has been compiling unsupported stuff, did we know about this one?

**Comment by sharvil** [comment]
AFAICT, there's also no way to iterate in reverse over a `ModuleList` with TorchScript. Running modules in reverse is necessary for normalizing flow-based architectures.

I've tried various approaches to running modules in reverse:
  - `for m in reversed(self.module_list)`
  - `for m in self.module_list[::-1]`
  - add an attribute that's a simple Python list of `Module` in addition to the `ModuleList`, i.e. `self.mod_list: List[Module] = [module1, module2, ...]`

None of these are accepted by TorchScript.

What's the ETA on implementing this feature request? In the meantime, is there a workaround I can use?

**Comment by lithces** [comment]
@sharvil 
for me, self.module_lists[::-1] works

**Comment by kimdwkimdw** [comment]
@sharvil Did you find any solution for this?

**Comment by sharvil** [comment]
We abandoned TorchScript since it hasn't seen any meaningful development in years and, in retrospect, turned out to be a liability.

If you're still on TorchScript and you have a small fixed-sized `ModuleList` (which is often the case), you can do something like:

```
for i in range(len(self.module_list) - 1, 0, -1):
  for j, module in enumerate(self.module_list):
    if i == j:
      # call into `module` here
      break
```

Just iterate over module indices in reverse and do a linear forward scan over the `module_list` to match on each index. It's O(n^2) but perfectly fine for small n.

**Comment by janeyx99** [comment]
TorchScript has been deprecated in favor of torch.export as of 2.11. https://pytorch.org/blog/pytorch-2-11-release-blog/ Closing this issue as we are not further developing on it--please ping if you believe this issue was erroneously closed.
